dagql/call/id.go

```diff
@@ -0,0 +1,383 @@
+package call
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/dagger/dagger/dagql/call/callpbv1"
+	"github.com/opencontainers/go-digest"
+	"github.com/vektah/gqlparser/v2/ast"
+	"github.com/zeebo/xxh3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+func New() *ID {
+	// we start with nil so there's always a nil parent at the bottom
+	return nil
+}
+
+/*
+ID represents a GraphQL value of a certain type, constructed by evaluating
+its contained pipeline. In other words, it represents a
+constructor-addressed value, which may be an object, an array, or a scalar
+value.
+
+It may be binary=>base64-encoded to be used as a GraphQL ID value for
+objects. Alternatively it may be stored in a database and referred to via an
+RFC-6920 ni://sha-256;... URI.
+
+This type wraps the underlying proto DAG+Call types in order to enforce immutability
+of its fields and give it a name more appropriate for how it's used in the
+context of dagql + the engine.
+
+IDs are immutable from the consumer's perspective. Rather than mutating an ID,
+methods on ID can be used to create a new ID on top of an existing (immutable)
+Base ID (e.g. Append, SelectNth, etc.).
+*/
+type ID struct {
+	pb *callpbv1.Call
+
+	// Wrappers around the various proto types in ID.pb
+	base   *ID
+	args   []*Argument
+	module *Module
+	typ    *Type
+}
+
+// The ID of the object that the field selection will be evaluated against.
+//
+// If nil, the root Query object is implied.
+func (id *ID) Base() *ID {
+	if id == nil {
+		return nil
+	}
+	return id.base
+}
+
+// The GraphQL type of the value.
+func (id *ID) Type() *Type {
+	return id.typ
+}
+
+// GraphQL field name.
+func (id *ID) Field() string {
+	return id.pb.Field
+}
+
+// GraphQL field arguments, always in alphabetical order.
+// NOTE: use with caution, any inplace writes to elements of the returned slice
+// can corrupt the ID
+func (id *ID) Args() []*Argument {
+	return id.args
+}
+
+// If the field returns a list, this is the index of the element to select.
+// Note that this defaults to zero, which means there is no selection of
+// an element in the list. Non-zero indexes are 1-based.
+func (id *ID) Nth() int64 {
+	return id.pb.Nth
+}
+
+// Tainted returns true if the ID contains any tainted selectors.
+// If true, this Selector is not reproducible.
+func (id *ID) IsTainted() bool {
+	if id == nil {
+		return false
+	}
+	if id.pb.Tainted {
+		return true
+	}
+	if id.base != nil {
+		return id.base.IsTainted()
+	}
+	return false
+}
+
+// The module that provides the implementation of the field, if any.
+func (id *ID) Module() *Module {
+	return id.module
+}
+
+// Digest returns the digest of the encoded ID. It does NOT canonicalize the ID
+// first.
+func (id *ID) Digest() digest.Digest {
+	if id == nil {
+		return ""
+	}
+	return digest.Digest(id.pb.Digest)
+}
+
+func (id *ID) Inputs() ([]digest.Digest, error) {
```

Comment:
NB: on my otel branch I changed this to include the receiver, but nothing uses this at the moment, and arguably the receiver is already accessible, so maybe it's helpful to have a distinction in case we want to treat argument + receiver differently in the UI. So I'm good with leaving this as-is.